### PR TITLE
GC volumeAttachment

### LINKED ISSUE #77324: Garbage collect VolumeAttachment objects to trigger detach
**What happened**:

*Background:* The attach/detach controller create `VolumeAttachment` objects for CSI Volumes to indicate that a volume needs to be attached to a node. It deletes the object to indicate that the volume should be detached.

*Problem:* If the Attach/Detach controller crashes, and in the meantime the pod using that volume is deleted, when the AD controller restarts, it won't know that the volume was previously attached and should be cleaned up.

**What you expected to happen**:

*Proposal:* We should have the AD controller periodically list all `VolumeAttachment` objects and for any that are not accounted for, update its internal actual/desired state caches so that the volume is gracefully detached.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:
/sig storage

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by cwdsuzhou**
**What type of PR is this?**

/kind cleanup
/kind feature

**What this PR does / why we need it**:
See https://github.com/kubernetes/kubernetes/issues/77324,  If the Attach/Detach controller crashes, and in the meantime the pod using that volume is deleted, when the AD controller restarts, it won't know that the volume was previously attached and should be cleaned up.
**Which issue(s) this PR fixes**:

Fixes #77324 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/controller/volume/attachdetach/attach_detach_controller.go [hunk 1]
```diff
@@ -358,18 +380,68 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
...
+		nodeName := types.NodeName(va.Spec.NodeName)
+		if va.Spec.Source.PersistentVolumeName == nil {
+			continue
+		}
+		var volumeName string
+		driverName := va.Spec.Attacher
+		pluginName := csi.CSIPluginName
+		pvName := *va.Spec.Source.PersistentVolumeName
+		pv, err := adc.pvLister.Get(pvName)
+		var uniqueVolumeName v1.UniqueVolumeName
```
Comment (cwdsuzhou) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go]: We need consider pv not existed

**Comment by cwdsuzhou** [comment]
ping @saad-ali  for review

File: pkg/controller/volume/attachdetach/attach_detach_controller.go [hunk 2]
```diff
@@ -358,18 +380,68 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
...
+		if err != nil && apierrors.IsNotFound(err) ||
+			pv == nil {
+			deleteError := adc.kubeClient.StorageV1().VolumeAttachments().Delete(va.Name, nil)
```
Comment (saad-ali) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go]: Deletion seems like a risky operation. Is it safe in all cases?

File: pkg/controller/volume/attachdetach/attach_detach_controller_test.go [hunk 3]
```diff
@@ -46,6 +46,7 @@ func Test_NewAttachDetachController_Positive(t *testing.T) {
 		informerFactory.Core().V1().PersistentVolumes(),
 		informerFactory.Storage().V1beta1().CSINodes(),
 		informerFactory.Storage().V1beta1().CSIDrivers(),
```
Comment (saad-ali) [inline pkg/controller/volume/attachdetach/attach_detach_controller_test.go]: Can you add a unit test for the new behavior?

**Review by saad-ali — COMMENTED** [review]
/assign @misterikkit

File: pkg/controller/volume/attachdetach/attach_detach_controller_test.go (hunk 3 above)
Comment (cwdsuzhou) [reply pkg/controller/volume/attachdetach/attach_detach_controller_test.go]: Sure, I would add some unit tests.

But before that, I should solve the conflicts first

File: pkg/controller/volume/attachdetach/attach_detach_controller.go (hunk 2 above)
Comment (cwdsuzhou) [reply pkg/controller/volume/attachdetach/attach_detach_controller.go]: I have added double check.
if pv does not exist in cache, try to get it from api-server?


wdyt?

**Comment by cwdsuzhou** [comment]
> @cwdsuzhou so sorry this appears to have fell off the radar. Are you still interested in working on this?

Thanks @msau42. I am interested in it, but I am very busy this week. I would like reopen this PR next week and work on this

**Pull request closed without merging**